scipy/stats/_stats_mstats_common.py

```diff
@@ -135,55 +142,53 @@ def theilslopes(y, x=None, alpha=0.95, method='separate'):
     if method not in ['joint', 'separate']:
         raise ValueError("method must be either 'joint' or 'separate'."
                          f"'{method}' is invalid.")
-    # We copy both x and y so we can use _find_repeats.
-    y = np.array(y, dtype=float, copy=True).ravel()
-    if x is None:
-        x = np.arange(len(y), dtype=float)
-    else:
-        x = np.array(x, dtype=float, copy=True).ravel()
-        if len(x) != len(y):
-            raise ValueError("Array shapes are incompatible for broadcasting.")
-    if len(x) < 2:
-        raise ValueError("`x` and `y` must have length at least 2.")
+
+    y, x = xp_promote(y, x, force_floating=True, xp=np)
+    x = np.arange(y.shape[-1], dtype=y.dtype) if x is None else x
+    y, x = np.broadcast_arrays(y, x)
 
     # Compute sorted slopes only when deltax > 0
-    deltax = x[:, np.newaxis] - x
-    deltay = y[:, np.newaxis] - y
-    slopes = deltay[deltax > 0] / deltax[deltax > 0]
```

Comment:
`deltax` and `deltay` are skew-symmetric with zeros on the diagonal, so `deltay/deltax` is symmetric with `nan`s on the diagonal. The main purpose of the `deltax > 0` mask is to eliminate all the duplicates and `nan`s before taking the median; we do the same thing using an upper triangular mask.

We are *also* supposed to ignore any other places where `deltax == 0`; we do that by replacing these elements with `nan` and using `nanmedian`. We can use `quantile` with `nan_policy='omit'` when we translate to array API.